### PR TITLE
User Groups Paywall

### PR DESCRIPTION
**Pull request by MrFlashAccount**
#### Tl;dr 
Closes: enso-org/cloud-v2#1200

This PR adds Initial paywall functionality inside enso and adds paywalls for the Users Group page in settings.

<details><summary>Demo Presentation</summary>
<p>

https://github.com/enso-org/enso/assets/61194245/b0b775d6-5946-4120-a422-f54a86a81719

</p>
</details> 

---

#### This Change:
What this change does in the larger context. Specific details to highlight for review:
1. Adds a new set of components for Paywall functionality
2. Closes the UsersGroup functionaly behind 2 paywalls - first is for personal/free acc, and the second is for team acc

#### Test Plan:
The only catch with testing is how to switch UI between different states, and I don't have an answer except switching between accounts or modifying the code to invert the logic.

---

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

**Comment by somebody1234** [comment]
mostly unrelated, but: i don't think we should let users on the Team plan manage user groups.
iirc the idea is, *everyone* is in the same user group, unconditionally. so the user group can't be deleted, and users can't be excluded from the group

File: app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import * as ariaComponents from '#/components/AriaComponents'
+
+import * as components from './components'
+import * as upgradeButton from './UpgradeButton'
+
+/**
+ * Props for a {@link PaywallDialog}.
+ */
+export type PaywallDialogProps = ariaComponents.DialogProps & {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialog.tsx]: use `extends` instead of an intersection type.

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx [hunk 2]
```diff
@@ -43,3 +43,17 @@ export const ALL_PATHS_REGEX = new RegExp(
...
+ * Build a Subscription URL for a given plan.
+ */
+export function getUpgradeURL(plan: string): string {
+  return SUBSCRIBE_PATH + '?plan=' + plan
+}
+
+/**
+ * Build a Subscription URL for contacting sales.
+ */
+export function getContactSalesURL(): string {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: is there a reason this should be a function?

**Comment by MrFlashAccount** [comment]
@somebody1234 We might want to move this discussion outside of the PR. Maybe move it to the related Issue? I guess we need opinions from @PabloBuchu and/or from @jdunkerley

File: app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialogButton.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import * as ariaComponents from '#/components/AriaComponents'
+
+import * as components from './components'
+import * as paywallDialog from './PaywallDialog'
+
+/**
+ * Props for a {@link PaywallDialogButton}.
+ */
+export type PaywallDialogButtonProps = Omit<ariaComponents.ButtonProps, 'variant'> & {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialogButton.tsx]: likewise, `extends` here (and for all the other components in this PR)

File: app/ide-desktop/eslint.config.js [hunk 4]
```diff
@@ -137,14 +133,6 @@ const RESTRICTED_SYNTAXES = [
         selector: `TSAsExpression:not(:has(TSTypeReference > Identifier[name=const]))`,
         message: 'Avoid `as T`. Consider using a type annotation instead.',
     },
-    {
```
Comment (MrFlashAccount) [inline app/ide-desktop/eslint.config.js:140]: This rule makes no sense with '[exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig/exactOptionalPropertyTypes.html)`.

File: app/ide-desktop/eslint.config.js [hunk 5]
```diff
@@ -461,11 +449,6 @@ export default [
                     selector: ':not(TSModuleDeclaration)[declare=true]',
                     message: 'No ambient declarations',
                 },
-                {
```
Comment (MrFlashAccount) [inline app/ide-desktop/eslint.config.js:464]: This rule makes no sense in case if we re-import default statements from barrel files

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: also i don't think this is the right place to put these functions, as they are not being used in `App.tsx`, nor are they related to `App.tsx`.

File: app/ide-desktop/eslint.config.js [hunk 6]
```diff
@@ -87,10 +87,6 @@ const RESTRICTED_SYNTAXES = [
         selector: `:matches(ImportDefaultSpecifier[local.name=/^${NAME}/i], ImportNamespaceSpecifier > Identifier[name=/^${NAME}/i])`,
         message: `Don't prefix modules with \`${NAME}\``,
     },
-    {
```
Comment (MrFlashAccount) [inline app/ide-desktop/eslint.config.js:90]: Removed it because creating an interface sometimes it too annoying

File: app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  const isEnterprise = level === billingHooks.PAYWALL_LEVELS.enterprise
+  const child =
+    children ?? (isEnterprise ? getText('contactSales') : getText('upgradeTo', levelLabel))
+
+  return (
+    <ariaComponents.Button
+      variant={variant}
+      size={size}
+      href={
+        isEnterprise ? appUtils.getContactSalesURL() : href ?? appUtils.getUpgradeURL(level.name)
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx:45]: - consider a single function instead of a ternary
- why are you using `level.name` here, but `level ===` in the comparison above?

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * Paywall configuration for different plans.
+ */
+
+import type * as text from '#/text'
+
+import * as backend from '#/services/Backend'
+
+/**
+ *
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts]: blank comment in various places in this file.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ */
+export const PAYWALL_FEATURES = {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:14]: prefer enum over an object.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ */
+export type PaywallLevelValue =
+  | (0 & { readonly name: PaywallLevelName; readonly label: text.TextId })
+  | (1 & { readonly name: PaywallLevelName; readonly label: text.TextId })
+  | (2 & { readonly name: PaywallLevelName; readonly label: text.TextId })
+  | (3 & { readonly name: PaywallLevelName; readonly label: text.TextId })
+
+export const PAYWALL_LEVELS: Record<PaywallLevelName, PaywallLevelValue> = {
+  free: Object.assign(0, { name: 'free', label: 'freePlanName' } as const),
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: is there a good reason to use `Number` objects here? why are the numbers not simply a property of the object?

File: app/ide-desktop/lib/dashboard/src/text/index.ts [hunk 11]
```diff
@@ -101,6 +101,10 @@ interface PlaceholderOverrides {
   readonly getDefaultVersionBackendError: [string]
 
   readonly subscribeSuccessSubtitle: [string]
+
+  readonly paywallAvailabilityLevel: [Plan: string]
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/text/index.ts]: parameter names should be `camelCase`, not `PascalCase`.

File: app/ide-desktop/lib/dashboard/src/components/styled/HorizontalMenuBar.tsx [hunk 12]
```diff
@@ -17,7 +17,7 @@ export default function HorizontalMenuBar(props: HorizontalMenuBarProps) {
   return (
     <FocusArea direction="horizontal">
       {innerProps => (
-        <div className="flex h-row gap-drive-bar" {...innerProps}>
+        <div className="flex gap-drive-bar" {...innerProps}>
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/styled/HorizontalMenuBar.tsx:20]: why the change in the class list?

File: app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallLock.tsx [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  const { getText } = textProvider.useText()
+
+  const { getFeature } = billingHooks.usePaywallFeatures()
+
+  const { level } = getFeature(feature)
+  const levelLabel = getText(level.label)
+
+  return (
+    <div className={tw.twMerge('flex w-full items-center gap-1 text-sm font-normal', className)}>
+      <SvgMask src={LockIcon} role="presentation" className="h-4 w-4" />
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallLock.tsx:36]: is the `role` prop needed here? `presentation` is the default `role` for an `img` without `alt`, no?

File: app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallButton.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+ * Props of the PaywallDialogButton component
+ */
+export type PaywallButtonProps = Omit<ariaComponents.ButtonProps, 'variant'> & {
+  readonly feature: billingHooks.PaywallFeatureName
+  readonly variant?: ariaComponents.ButtonProps['variant']
+  readonly iconOnly?: boolean
+}
+
+/**
+ * PaywallDialogButton component
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallButton.tsx]: incorrect name + docs don't describe the component

File: app/ide-desktop/eslint.config.js (hunk 6 above)
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:90]: see review comments

File: app/ide-desktop/eslint.config.js (hunk 4 above)
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:140]: `exactOptionalPropertyTypes` is not related to this rule. if anything, `exactOptionalPropertyTypes` *removes* `undefined` from the types of optional properties (unless explicitly declared as `| undefined`), meaning that this rule shouldn't cause any problem at all.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 10 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: Because plans are Inclusive - all features from a level below are included in the levels above. 
So if we want to check that the feature is available, we can simply compare 2 numbers - the user's level and the feature level.

File: app/ide-desktop/lib/dashboard/src/components/styled/HorizontalMenuBar.tsx (hunk 12 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/styled/HorizontalMenuBar.tsx:20]: Because setting explicit height is a bad practice in general

File: app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallLock.tsx (hunk 13 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/components/PaywallLock.tsx:36]: If Img doesn't have an alt, screenreaders pronounce the src content. We can either set alt="" or role="presentation"

File: app/ide-desktop/eslint.config.js (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/eslint.config.js:140]: In case I want to allow passing smth: undefined,  I need to specify it explicitly, but by specifying it explicitly, I'm making eslint angry.

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: They're here because they're related to Paths(and all paths are specified here). Eventually, we'll move the URLs closer to their pages and the function will follow.

File: app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialogButton.tsx (hunk 3 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/PaywallDialogButton.tsx]: Here it won't work because `ariaComponents.ButtonProps` is a complicated Discriminative Union

File: app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx (hunk 7 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx:45]: Because `level.name` is a string? and level is a number? Numbers are easier to compare.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 9 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:14]: Why, though? enums have no advantages over regular objects.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 10 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: we can do that with a property as well. `user.plan.level < plan.level`

File: app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx (hunk 7 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx:45]: right, so the question is, why does `getUpgradeURL` not accept `level`, but instead requires `level.name` specifically?

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: > all paths are specified here

correction: all paths *used in `App.tsx`* are specified here. the paths are (or at least, should be) used nowhere outside of `App.tsx`

File: app/ide-desktop/eslint.config.js (hunk 4 above)
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:140]: @MrFlashAccount but the point of `undefined` is that you *don't* need to pass it explicitly. if you need to omit something, simply don't specify the property at all

File: app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx (hunk 7 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx:45]: Because it shouldn't know anything about paywalls. It accepts a string and puts it in a `plan` search param.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 10 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: I see no arguments why `user.plan.level < plan.level` is better than `level < userLevel`

File: app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx (hunk 7 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/Paywall/UpgradeButton.tsx:45]: this function is a URL string builder. not a paywall-related thing

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: > correction: all paths used in App.tsx are specified here. the paths are (or at least, should be) used nowhere outside of App.tsx

Since we don't specify paths outside of the `App.tsx` this correction makes no change

File: app/ide-desktop/eslint.config.js (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/eslint.config.js:140]: Often we use the `object destruction` pattern and checking if a property is not undefined before passing it is tedious. 

See: https://github.com/enso-org/enso/blob/develop/app/ide-desktop/lib/dashboard/src/components/AriaComponents/Form/Form.tsx#L55

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: `getContactSalesURL` is used only in `getComponentForPlan.tsx` and `UpgradeButton.tsx`, neither of which are `App.tsx` - unless i'm misunderstanding something.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 9 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:14]: similarly: regular objects have no advantages over enums.
(plus you have extra boilerplate that you need to remember to add:
- `as const`
- you need a separate type definition to create the type (enums have the type built-in)
- enums have extra type safety (they are nominally typed)

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 10 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: it's not! but i'm really, *really* unsure about passing around a `Number` object, just so the code looks a tiny bit cleaner in one specific situation.

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 9 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:14]: > similarly: regular objects have no advantages over enums.

It's not true, because objects are:
1. Js-compatible
2. Do not emit extra code in runtime
3. Do not have pitfalls enums have

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:14]: > enums have extra type safety (they are nominally typed)

function some(thing: MyEnum) {}
is not safer than
function some(thing: MyObjType) {}

More details are here: https://www.youtube.com/watch?v=jjMbPt_H3RQ
(I guess it's the best video that describes my experience with enums)

File: app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts (hunk 10 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: I my experience checking things like 'level === PaywallLevels.enterprise' or 'level > PaywallLevels.enterpriseEntry' is more than just a `one specific case`. 
Also, since `level` is a primitive type, we don't care if references don't match. 
CleanShot 2024-05-22 at 19 23 12

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/hooks/billing/FeaturesConfiguration.ts:45]: I agree this looks __strange__ and unusual, but it has advantages and doesn't violate the agreement we have in javascript, like `do not mutate prototypes`

File: app/ide-desktop/lib/dashboard/src/appUtils.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: We keep this function with the URL just because we don't want to forget it when we refactor the Router-related stuff and put to it's own place under `/pages`

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/appUtils.tsx:57]: But if you have a better option - I'll comply

**Review by somebody1234 — CHANGES_REQUESTED** [review]
buttons in settings sidebar no longer have lowered opacity when inactive (they are supposed to).

File: app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx [hunk 15]
```diff
@@ -133,19 +134,22 @@ export default function SettingsSidebar(props: SettingsSidebarProps) {
...
-                  icon={tab.icon}
-                  label={tab.name}
-                  active={tab.settingsTab === settingsTab}
-                  onPress={() => {
-                    setSettingsTab(tab.settingsTab)
-                  }}
-                />
-              ))}
+
+              <ariaComponents.ButtonGroup gap="xxsmall" direction="column" align="start">
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx:138]: please try to avoid changing layout from the figma design. (the figma design does not have a gap)

**Comment by MrFlashAccount** [comment]
> buttons in the settings sidebar no longer have lowered opacity when inactive (they are supposed to).

It's intended, we decided to make the buttons look like that with @PabloBuchu because currently, the inactive items look disabled.

File: app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx (hunk 15 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx:138]: Gaps are intended because, even though it's not described in the design if we remove the gaps, menu items will look like they stuck together(see screenshot). Alternatively, we can use a different hover for elements.

CleanShot 2024-05-23 at 13 58 45

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx:138]: fair. in that case, we should probably reduce the size of the menu entries and add vertical padding so that the overall height in pixels is still the same? (32 * entries.length)

**Review by somebody1234 — COMMENTED** [review]
requested changes have been resolved

File: app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx (hunk 15 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/SettingsSidebar.tsx:138]: alternatively reduce by just 1px or 0.5px to make the gaps consistent with those ofnthe assets table rows

**Comment by MrFlashAccount** [comment]
> requested changes have been resolved

Can I ask to approve the PR then?

**Review by somebody1234 — APPROVED** [review]
github moment

**Comment by somebody1234** [comment]
(remember that QA is not yet done though)

**Comment by MrFlashAccount** [comment]
Closing this PR in favor of https://github.com/enso-org/enso/pull/10087
